modules/sitemaps/sitemap-librarian.php

```diff
@@ -0,0 +1,593 @@
+	 */
+	public function query_latest_approved_comment_time_on_post( $post_id ) {
+		global $wpdb;
+
+		return $wpdb->get_var(
+			$wpdb->prepare(
+				"SELECT MAX(comment_date_gmt)
+					FROM $wpdb->comments
+					WHERE comment_post_ID = %d AND comment_approved = '1' AND comment_type=''",
+				$post_id
+			)
+		);
+	}
+
+	/**
+	 * Retrieve an array of image posts sorted by ID.
+	 *
+	 * More precisely, returns the smallest $num_posts image posts
+	 * (measured by ID) which are larger than $from_id.
+	 *
+	 * @access public
+	 * @since 4.6.0
+	 *
+	 * @param int $from_id Greatest lower bound of retrieved image post IDs.
+	 * @param int $num_posts Largest number of image posts to retrieve.
+	 *
+	 * @return array The posts.
+	 */
+	public function query_images_after_id( $from_id, $num_posts ) {
+		global $wpdb;
+
+		return $wpdb->get_results(
+			$wpdb->prepare(
+				"SELECT *
+					FROM $wpdb->posts
+					WHERE post_type='attachment'
+						AND post_mime_type IN ('image/jpeg','image/png','image/gif')
+						AND ID>%d
+					ORDER BY ID ASC
+					LIMIT %d;",
+				$from_id,
+				$num_posts
+			)
+		); // WPCS: db call ok; no-cache ok.
+	}
+
+	/**
+	 * Retrieve an array of video posts sorted by ID.
+	 *
+	 * More precisely, returns the smallest $num_posts video posts
+	 * (measured by ID) which are larger than $from_id.
+	 *
+	 * @access public
+	 * @since 4.6.0
+	 *
+	 * @param int $from_id Greatest lower bound of retrieved video post IDs.
+	 * @param int $num_posts Largest number of video posts to retrieve.
+	 *
+	 * @return array The posts.
+	 */
+	public function query_videos_after_id( $from_id, $num_posts ) {
+		global $wpdb;
+
+		return $wpdb->get_results(
+			$wpdb->prepare(
+				"SELECT *
+					FROM $wpdb->posts
+					WHERE post_type='attachment'
+						AND post_mime_type IN ('video/mpeg','video/wmv','video/mov')
+						AND ID>%d
+					ORDER BY ID ASC
+					LIMIT %d;",
+				$from_id,
+				$num_posts
+			)
+		); // WPCS: db call ok; no-cache ok.
+	}
+
+	/**
+	 * Retrieve an array of published posts from the last 2 days.
+	 *
+	 * @access public
+	 * @since 4.6.0
+	 *
+	 * @param int $num_posts Largest number of posts to retrieve.
+	 *
+	 * @return array The posts.
+	 */
+	public function query_most_recent_posts( $num_posts ) {
+		global $wpdb;
+
+		$two_days_ago = date( 'Y-m-d', strtotime( '-2 days' ) );
+
+		/**
+		 * Filter post types to be included in news sitemap.
+		 *
+		 * @module sitemaps
+		 *
+		 * @since 3.9.0
+		 *
+		 * @param array $post_types Array with post types to include in news sitemap.
+		 */
+		$post_types = apply_filters(
+			'jetpack_sitemap_news_sitemap_post_types',
+			array( 'page', 'post' )
+		);
+
+		foreach ( (array) $post_types as $i => $post_type ) {
+			$post_types[ $i ] = $wpdb->prepare( '%s', $post_type );
+		}
+
+		$post_types_list = join( ',', $post_types );
+
+		return $wpdb->get_results(
```

Comment:
`get_posts()` is a nice interface for the most common sorts of queries that removes the burden of constructing safe SQL statements. It probably does other magic too that I don't know about.

One thing I couldn't get it to do was include an `>` clause for post IDs. Which makes sense, because reasonable people don't care whether one post ID is bigger than another, just that they are different. But this plugin treats post IDs like array indices, so we need that. That's the reason for the raw queries.

Except for the news query here; that one probably can be replaced by a call to `get_posts()`.

As for caching, that's a good point. A malicious user could cause a bunch of db traffic by requesting the news sitemap over and over. But then caching makes us lose the freshness of the news sitemap; maybe it's possible to flush the cache whenever a post is published. I'll check this out in more depth.